src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Transmission/OtlpExporterPersistentStorageTransmissionHandler.cs

```diff
@@ -0,0 +1,190 @@
+                Debug.Fail("Unexpected request type encountered");
+                data = null;
+            }
+
+            if (data != null)
+            {
+                return this.persistentBlobProvider.TryCreateBlob(data, out _);
+            }
+        }
+
+        return false;
+    }
+
+    protected override void OnShutdown(int timeoutMilliseconds)
+    {
+        var sw = timeoutMilliseconds == Timeout.Infinite ? null : Stopwatch.StartNew();
+
+        try
+        {
+            this.shutdownEvent.Set();
+        }
+        catch (ObjectDisposedException)
+        {
+            // Dispose was called before shutdown.
+        }
+
+        this.thread.Join(timeoutMilliseconds);
+
+        if (sw != null)
+        {
+            var timeout = timeoutMilliseconds - sw.ElapsedMilliseconds;
+
+            base.OnShutdown((int)Math.Max(timeout, 0));
+        }
+        else
+        {
+            base.OnShutdown(timeoutMilliseconds);
+        }
+    }
+
+    protected override void Dispose(bool disposing)
+    {
+        if (!this.disposed)
+        {
+            if (disposing)
+            {
+                this.shutdownEvent.Dispose();
+                this.exportEvent.Dispose();
+                this.dataExportNotification.Dispose();
+                (this.persistentBlobProvider as IDisposable)?.Dispose();
+            }
+
+            this.disposed = true;
+        }
+    }
+
+    private void RetryStoredRequests()
+    {
+        var handles = new WaitHandle[] { this.shutdownEvent, this.exportEvent };
+        while (true)
+        {
+            try
+            {
+                var index = WaitHandle.WaitAny(handles, RetryIntervalInMilliseconds);
+                if (index == 0)
+                {
+                    // Shutdown signaled
+                    break;
+                }
+
+                int fileCount = 0;
+
+                // TODO: Run maintenance job.
+                // Transmit 10 files at a time.
+                while (fileCount < 10 && !this.shutdownEvent.WaitOne(0))
+                {
+                    if (!this.persistentBlobProvider.TryGetBlob(out var blob))
+                    {
+                        break;
+                    }
+
+                    if (blob.TryLease((int)this.TimeoutMilliseconds) && blob.TryRead(out var data))
+                    {
+                        var deadlineUtc = DateTime.UtcNow.AddMilliseconds(this.TimeoutMilliseconds);
+                        var request = this.requestFactory.Invoke(data);
+                        if (this.TryRetryRequest(request, deadlineUtc, out var response) || !RetryHelper.ShouldRetryRequest(request, response, OtlpRetry.InitialBackoffMilliseconds, out var retryInfo))
+                        {
+                            blob.TryDelete();
+                        }
+                        else
+                        {
+                            // Extend the lease period so that it is retried before the intended retry interval.
```

Comment:
```suggestion
                            // Extend the lease period so that it is not retried before the intended retry interval.
```

Reply:
😆 I removed it. Added TODO instead, wanted to check if I can add a test as well for it.